In [1]:
import graphlab

In [2]:
graphlab.canvas.set_target('ipynb')

In [3]:
## 1. read the data file

In [4]:
people = graphlab.SFrame('people_wiki.gl/')
people.head()

[INFO] graphlab.cython.cy_server: GraphLab Create v2.1 started. Logging: /tmp/graphlab_server_1538452475.log


This non-commercial license of GraphLab Create for academic use is assigned to r38411@yahoo.com and will expire on August 31, 2019.


Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [5]:
## 2. build a word count

In [8]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

In [9]:
## 3. build a tfidf 

In [10]:
people['tfidf'] = graphlab.text_analytics.tf_idf(people['word_count'])
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'selection': 3.8365785530... |
| {'precise': 6.443200606955... |
| {'just': 2.700729968710864... |
| {'all': 1.6431112434912472... |
| {'they': 1.899340117819389... |
| {'currently': 1.6370889691... |
| {'exclusive': 10.455187230... |
| {'taxi': 6.052021456094502... |
| {'houston': 3.935505942157... |
| {'phenomenon': 5.750053426... |
+-------------------------------+
[10 rows x 5 columns]

In [11]:
## 4. retrieve elton john

In [12]:
elton = people[people['name'] == 'Elton John']
elton

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'all': 1, 'least': 1, 'pr... | {'all': 1.6431112434912472... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [13]:
## 5. sort the elton word_count and get the top 3

In [15]:
elton[['word_count']].stack('word_count', new_column_name=['word','count']).sort('count', ascending=False)

Columns:
	word	str
	count	int

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   27  |
|   in  |   18  |
|  and  |   15  |
|   of  |   13  |
|   a   |   10  |
|  has  |   9   |
|  john |   7   |
|   he  |   7   |
|   on  |   6   |
| award |   5   |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [16]:
elton[['tfidf']].stack('tfidf', new_column_name=['tf_word','tf_count']).sort('tf_count',ascending=False)

Columns:
	tf_word	str
	tf_count	float

Rows: 255

Data:
+---------------+---------------+
|    tf_word    |    tf_count   |
+---------------+---------------+
|    furnish    |  18.38947184  |
|     elton     |  17.48232027  |
|   billboard   | 17.3036809575 |
|      john     | 13.9393127924 |
|  songwriters  |  11.250406447 |
| tonightcandle | 10.9864953892 |
|  overallelton | 10.9864953892 |
|    19702000   | 10.2933482087 |
|   fivedecade  | 10.2933482087 |
|      aids     |  10.262846934 |
+---------------+---------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Question 1: Elton
# 3 top word counts: the(27), in(18), and(15)
# 3 top tfidf counts: furnish(18.389), elton(17.482), billboard(17.303), john(13.93), songwriters(11.25)

In [18]:
beckham = people[people['name'] == 'Victoria Beckham']
beckham

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------------+
|              URI              |       name       |
+-------------------------------+------------------+
| <http://dbpedia.org/resour... | Victoria Beckham |
+-------------------------------+------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| victoria caroline beckham ... | {'millionin': 1, 'saying':... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'millionin': 7.7283988512... |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [19]:
paul = people[people['name'] == 'Paul McCartney']
paul

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+----------------+
|              URI              |      name      |
+-------------------------------+----------------+
| <http://dbpedia.org/resour... | Paul McCartney |
+-------------------------------+----------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| sir james paul mccartney m... | {'all': 1, 'gold': 1, 'ove... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'all': 1.6431112434912472... |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [20]:
## find out who is closer to elton john: paul or victoria

In [23]:
john_paul = graphlab.distances.cosine(elton['tfidf'][0], paul['tfidf'][0])
john_vict = graphlab.distances.cosine(elton['tfidf'][0], beckham['tfidf'][0])
print 'john_paul = ', john_paul, ' , john_victoria = ', john_vict

john_paul =  0.825031002922  , john_victoria =  0.956700637666


# Question 2: 
# John-Paul cosine distance = 0.825, John-Beckham cosine distance = 0.9567
# Elton John is closer to Paul McCarthy than to Victoria Beckham

In [25]:
## 1. build 2 models based on: word_count and tfidf

In [32]:
knn_word_count_model = graphlab.nearest_neighbors.create(people, label='name',
                                                        features=['word_count'],
                                                        distance='cosine')
knn_tfidf_model = graphlab.nearest_neighbors.create(people, label='name',
                                                   features=['tfidf'],
                                                   distance='cosine')
knn_tfidf_model

Starting brute force nearest neighbors model training.

Starting brute force nearest neighbors model training.

Class                          : NearestNeighborsModel

Attributes
----------
Method                         : brute_force
Number of distance components  : 1
Number of examples             : 59071
Number of feature columns      : 1
Number of unpacked features    : 547979
Total training time (seconds)  : 2.9679

In [29]:
## 2. now find out the nearest neighbors of Elton John based on above models

In [33]:
knn_word_count_model.query(elton)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 7.174ms      |

| Done         |         | 100         | 197.104ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-------------------+------+
| query_label |  reference_label  |      distance     | rank |
+-------------+-------------------+-------------------+------+
|      0      |     Elton John    | 2.22044604925e-16 |  1   |
|      0      |   Cliff Richard   |   0.16142415259   |  2   |
|      0      |   Sandro Petrone  |   0.16822542751   |  3   |
|      0      |    Rod Stewart    |   0.168327165587  |  4   |
|      0      | Malachi O'Doherty |   0.177315545979  |  5   |
+-------------+-------------------+-------------------+------+
[5 rows x 4 columns]

In [34]:
knn_tfidf_model.query(elton)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 8.351ms      |

| Done         |         | 100         | 229.084ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+--------------------+------+
| query_label | reference_label  |      distance      | rank |
+-------------+------------------+--------------------+------+
|      0      |    Elton John    | -2.22044604925e-16 |  1   |
|      0      |   Rod Stewart    |   0.717219667893   |  2   |
|      0      |  George Michael  |   0.747600998969   |  3   |
|      0      | Sting (musician) |   0.747671954431   |  4   |
|      0      |   Phil Collins   |   0.75119324879    |  5   |
+-------------+------------------+--------------------+------+
[5 rows x 4 columns]

# Question 3: nearest neighbors
# Based on word_count_model, the closest neighbor of Elton John is: Cliff Richard(0.1614), Sandro Petrone(0.161),
#                                  Rod Steward(0.1683)
# Based on tfidf_model, the closest neighbor of ELton John is: Rod Steward(0.7172), George Michael(0.7476),
#                                  Sting(0.7476)

In [37]:
print 'word_count_model:', knn_word_count_model.query(beckham)
print 'tfidf_model', knn_tfidf_model.query(beckham)

word_count_model:

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 6.345ms      |

| Done         |         | 100         | 154.666ms    |

+--------------+---------+-------------+--------------+

+-------------+--------------------------+--------------------+------+
| query_label |     reference_label      |      distance      | rank |
+-------------+--------------------------+--------------------+------+
|      0      |     Victoria Beckham     | -2.22044604925e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |   0.207307036115   |  2   |
|      0      |      Adrienne Corri      |   0.214509782788   |  3   |
|      0      |     Beverly Jane Fry     |   0.217466468741   |  4   |
|      0      |      Raman Mundair       |   0.217695474992   |  5   |
+-------------+--------------------------+--------------------+------+
[5 rows x 4 columns]

tfidf_model

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 8.897ms      |

| Done         |         | 100         | 178.154ms    |

+--------------+---------+-------------+--------------+

+-------------+---------------------+-------------------+------+
| query_label |   reference_label   |      distance     | rank |
+-------------+---------------------+-------------------+------+
|      0      |   Victoria Beckham  | 1.11022302463e-16 |  1   |
|      0      |    David Beckham    |   0.548169610263  |  2   |
|      0      | Stephen Dow Beckham |   0.784986706828  |  3   |
|      0      |        Mel B        |   0.809585523409  |  4   |
|      0      |    Caroline Rush    |   0.819826422919  |  5   |
+-------------+---------------------+-------------------+------+
[5 rows x 4 columns]



# Question 3b:
# Based on word_count model, Beckham's closest neighbors are: Mary Fitzgerald(0.2073), 
#            Adriene Corri(0.2145), Beverly Jane Fry(0.217)
# Based on tfidf_model, Beckham's closest neighbors are:
#           David Beckham(0.548), Stephen Beckham(0.784), Mel B(0.809), Caroline Rush(0.8198)